{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Week 7 Assignment\n",
    "\n",
    "_MkKinney 6.1_\n",
    "\n",
    "This week has been all about getting information off the internet both in structured data formats (CSV, JSON, etc) as well as HTML.  For these exercises, we're going to use two practical examples of fetching data from web pages to show how to use Pandas and BeautifulSoup to extract structured information from the web.\n",
    "\n",
    "---\n",
    "---"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 33.1 Parsing a list in HTML\n",
    "\n",
    "Go to the Banner Health Price Transparency Page: https://www.bannerhealth.com/patients/billing/pricing-resources/hospital-price-transparency\n",
    "\n",
    "Notice that there is a list of hospitals and the city they are in.  We want to parse the underlying HTML to create a list of all the hospitals along with which city they're in.\n",
    "\n",
    "```json\n",
    "[\n",
    "    [\"Banner - University Medical Center Phoenix\", \"Arizona\"],\n",
    "    [\"Banner - University Medical Center South \", \"Arizona\"],\n",
    "    ...\n",
    "]\n",
    "```\n",
    "\n",
    "To examine the underlying HTML code, you can use Chrome, right-click, and choose **Inspect**.\n",
    "\n",
    "For reference, the documentation for BeautifulSoup is here: https://www.crummy.com/software/BeautifulSoup/bs4/doc/\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "from bs4 import BeautifulSoup\n",
    "import requests\n",
    "headers = { \"User-Agent\": \"Mozilla/5.0 (Macintosh; Intel Mac OS X 11_2_3) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/89.0.4389.90 Safari/537.36\" }\n",
    "\n",
    "response = requests.get('https://www.bannerhealth.com/patients/billing/pricing-resources/hospital-price-transparency', headers=headers)\n",
    "soup = BeautifulSoup(response.text, 'html.parser')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Arizona Banner - University Medical Center Phoenix\n",
      "Arizona Banner - University Medical Center South \n",
      "Arizona Banner - University Medical Center Tucson\n",
      "Arizona Banner Baywood Medical Center \n",
      "Arizona Banner Behavioral Health Hospital\n",
      "Arizona Banner Boswell Medical Center\n",
      "Arizona Banner Casa Grande Medical Center\n",
      "Arizona Banner Del E. Webb Medical Center\n",
      "Arizona Banner Desert Medical Center/Cardon Children's Medical Center  \n",
      "Arizona Banner Estrella Medical Center\n",
      "Arizona Banner Gateway Medical Center/Banner MD Anderson Cancer Center\n",
      "Arizona Banner Goldfield Medical Center  \n",
      "Arizona Banner Heart Hospital\n",
      "Arizona Banner Ironwood Medical Center\n",
      "Arizona Banner Ocotillo Medical Center\n",
      "Arizona Banner Payson Medical Center\n",
      "Arizona Banner Thunderbird Medical Center\n",
      "Arizona Page Hospital\n",
      "California Banner Lassen Medical Center\n",
      "Colorado Banner Fort Collins Medical Center\n",
      "Colorado McKee Medical Center\n",
      "Colorado North Colorado Medical Center\n",
      "Colorado Sterling Regional Medical Center\n",
      "Colorado East Morgan County Hospital\n",
      "Nebraska Ogallala Community Hospital\n",
      "Nevada Banner Churchill Community Hospital\n",
      "Wyoming Banner Wyoming Medical Center\n",
      "Wyoming Community Hospital\n",
      "Wyoming Washakie Medical Center\n",
      "Wyoming Platte County Memorial Hospital\n",
      "State-Specific Regulations Arizona\n",
      "State-Specific Regulations California\n",
      "State-Specific Regulations Colorado\n",
      "State-Specific Regulations Nebraska\n",
      "State-Specific Regulations Nevada\n",
      "State-Specific Regulations Wyoming\n"
     ]
    }
   ],
   "source": [
    "div = soup.find_all('div', {\"class\":\"col-md-8\"})[0]\n",
    "for hospital_list in div.find_all('ul'):\n",
    "    state = hospital_list.previous_sibling.previous_sibling.string\n",
    "    for hospital in hospital_list.find_all('li'):\n",
    "        print(state, hospital.text)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "---\n",
    "\n",
    "### 33.2 Using Pandas to Read Tables\n",
    "\n",
    "\n",
    "Pandas documentation for loading data https://pandas.pydata.org/pandas-docs/version/0.23.4/api.html#input-output\n",
    "\n",
    "Pandas documentation for describing the shape of data https://pandas.pydata.org/pandas-docs/version/0.23.4/generated/pandas.DataFrame.shape.html"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "5"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tables = pd.read_html('https://en.wikipedia.org/wiki/Diagnosis-related_group')\n",
    "len(tables)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "**************TABLE 0\n",
      "    0                                                  1\n",
      "0 NaN  This article has multiple issues. Please help ...\n",
      "1 NaN  This article needs to be updated. Please updat...\n",
      "2 NaN  This article needs additional citations for ve...\n",
      "**************TABLE 1\n",
      "    0                                                  1\n",
      "0 NaN  This article needs to be updated. Please updat...\n",
      "**************TABLE 2\n",
      "    0                                                  1\n",
      "0 NaN  This article needs additional citations for ve...\n",
      "**************TABLE 3\n",
      "   Hypothetical patient at Generic Hospital in San Francisco, CA, DRG 482, HIP & FEMUR PROCEDURES EXCEPT MAJOR JOINT W/O CC/MCC (2001)[15]:8  \\\n",
      "0                                         Description                                                                                          \n",
      "1                              Average length of stay                                                                                          \n",
      "2                      Large urban labor-related rate                                                                                          \n",
      "3                       Large urban non-labor-related                                                                                          \n",
      "4                                          Wage index                                                                                          \n",
      "5   Standard Federal Rate: labor * wage index + no...                                                                                          \n",
      "6                     DRG relative weight (RW) factor                                                                                          \n",
      "7    Weighted payment: Standard Federal Rate * DRG RW                                                                                          \n",
      "8                Disproportionate Share Payment (DSH)                                                                                          \n",
      "9                    Indirect medical education (IME)                                                                                          \n",
      "10                   Total cost outlier reimbursement                                                                                          \n",
      "11  Total operating payment: Weighted payment * (1...                                                                                          \n",
      "\n",
      "   Hypothetical patient at Generic Hospital in San Francisco, CA, DRG 482, HIP & FEMUR PROCEDURES EXCEPT MAJOR JOINT W/O CC/MCC (2001)[15]:8.1  \n",
      "0                                               Value                                                                                           \n",
      "1                                             3.8[16]                                                                                           \n",
      "2                                           $2,809.18                                                                                           \n",
      "3                                           $1,141.85                                                                                           \n",
      "4                                              1.4193                                                                                           \n",
      "5                                           $5,128.92                                                                                           \n",
      "6                                              1.8128                                                                                           \n",
      "7                                           $9,297.71                                                                                           \n",
      "8                                              0.1413                                                                                           \n",
      "9                                              0.0744                                                                                           \n",
      "10                                                 $0                                                                                           \n",
      "11                                         $11,303.23                                                                                           \n",
      "**************TABLE 4\n",
      "     Name  Version       Start date  \\\n",
      "0  MS-DRG       25  October 1, 2007   \n",
      "1  MS-DRG       26  October 1, 2008   \n",
      "2  MS-DRG       27  October 1, 2009   \n",
      "3  MS-DRG       31  October 1, 2013   \n",
      "4  MS-DRG       32  October 1, 2014   \n",
      "5  MS-DRG       33  October 1, 2015   \n",
      "6  MS-DRG       34  October 1, 2016   \n",
      "7  MS-DRG       35  October 1, 2017   \n",
      "\n",
      "                                               Notes  \n",
      "0  Group numbers resequenced, so that for instanc...  \n",
      "1  One main change: implementation of Hospital Ac...  \n",
      "2  Changes involved are mainly related to Influen...  \n",
      "3                                                NaN  \n",
      "4                                                NaN  \n",
      "5            Convert from ICD-9-CM to ICD-10-CM.[17]  \n",
      "6  Address ICD-10 replication issues introduced i...  \n",
      "7  MS-DRGs 984 through 986 deleted and reassigned...  \n"
     ]
    }
   ],
   "source": [
    "for index,table in enumerate(tables):\n",
    "    print(\"**************TABLE {}\".format(index))\n",
    "    print(table)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Name</th>\n",
       "      <th>Version</th>\n",
       "      <th>Start date</th>\n",
       "      <th>Notes</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>MS-DRG</td>\n",
       "      <td>25</td>\n",
       "      <td>October 1, 2007</td>\n",
       "      <td>Group numbers resequenced, so that for instanc...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>MS-DRG</td>\n",
       "      <td>26</td>\n",
       "      <td>October 1, 2008</td>\n",
       "      <td>One main change: implementation of Hospital Ac...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>MS-DRG</td>\n",
       "      <td>27</td>\n",
       "      <td>October 1, 2009</td>\n",
       "      <td>Changes involved are mainly related to Influen...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>MS-DRG</td>\n",
       "      <td>31</td>\n",
       "      <td>October 1, 2013</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>MS-DRG</td>\n",
       "      <td>32</td>\n",
       "      <td>October 1, 2014</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>MS-DRG</td>\n",
       "      <td>33</td>\n",
       "      <td>October 1, 2015</td>\n",
       "      <td>Convert from ICD-9-CM to ICD-10-CM.[17]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>MS-DRG</td>\n",
       "      <td>34</td>\n",
       "      <td>October 1, 2016</td>\n",
       "      <td>Address ICD-10 replication issues introduced i...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>MS-DRG</td>\n",
       "      <td>35</td>\n",
       "      <td>October 1, 2017</td>\n",
       "      <td>MS-DRGs 984 through 986 deleted and reassigned...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     Name  Version       Start date  \\\n",
       "0  MS-DRG       25  October 1, 2007   \n",
       "1  MS-DRG       26  October 1, 2008   \n",
       "2  MS-DRG       27  October 1, 2009   \n",
       "3  MS-DRG       31  October 1, 2013   \n",
       "4  MS-DRG       32  October 1, 2014   \n",
       "5  MS-DRG       33  October 1, 2015   \n",
       "6  MS-DRG       34  October 1, 2016   \n",
       "7  MS-DRG       35  October 1, 2017   \n",
       "\n",
       "                                               Notes  \n",
       "0  Group numbers resequenced, so that for instanc...  \n",
       "1  One main change: implementation of Hospital Ac...  \n",
       "2  Changes involved are mainly related to Influen...  \n",
       "3                                                NaN  \n",
       "4                                                NaN  \n",
       "5            Convert from ICD-9-CM to ICD-10-CM.[17]  \n",
       "6  Address ICD-10 replication issues introduced i...  \n",
       "7  MS-DRGs 984 through 986 deleted and reassigned...  "
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "drgs = tables[4]\n",
    "drgs"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "---\n",
    "\n",
    "### 33.3 Find Something of Your Own\n",
    "\n",
    "Do some web searches and find an HTML page with some data that is interesting to something you're studying.  You can extract and parse that information using either BeautifulSoup or Pandas.  If you're using Pandas, then do something interesting to format and structure your data.  If you're using BeautifulSoup, you'll just need to do the work of parsing the data out of HTML -- that's hard enough!\n",
    "\n",
    "You don't need to build this as a function.  Just use notebook cells as I've done above.  You will be graded based on _style_.  Use variable names that make sense for your problem / solution. Cleanup anything you don't need before you submit your work."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Your Code Here"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "4"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tables = pd.read_html('https://en.wikipedia.org/wiki/CRISPR_gene_editing')\n",
    "len(tables)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "**************TABLE 0\n",
      "         Component                                           Function\n",
      "0            crRNA  Contains the guide RNA that locates the correc...\n",
      "1         tracrRNA        Binds to crRNA and forms an active complex.\n",
      "2            sgRNA  Single-guide RNAs are a combined RNA consistin...\n",
      "3             Cas9  An enzyme whose active form is able to modify ...\n",
      "4  Repair template  DNA molecule used as a template in the host ce...\n",
      "**************TABLE 1\n",
      "                                             Library  \\\n",
      "0               Bassik Mouse CRISPR Knockout Library   \n",
      "1  Mouse Tumor Suppressor Gene CRISPR Knockout Li...   \n",
      "2                     Brie mouse genome-wide library   \n",
      "3               Bassik Human CRISPR Knockout Library   \n",
      "4                 Brunello human genome-wide library   \n",
      "5  Mini-human AsCpf1-based Human Genome-wide Knoc...   \n",
      "\n",
      "                                        ID Species               PI  \\\n",
      "0                  1000000121 — 1000000130   Mouse           Bassik   \n",
      "1  113584 EFS backbone 113585 TBG backbone   Mouse             Chen   \n",
      "2      73632 (1 plasmid) 73633 (2 plasmid)   Mouse  Doench and Root   \n",
      "3                          101926 — 101934   Human           Bassik   \n",
      "4      73179 (1 plasmid) 73178 (2 plasmid)   Human  Doench and Root   \n",
      "5                                   130630   Human          Draetta   \n",
      "\n",
      "              Genes targeted gRNAs per gene    Total gRNAs  \n",
      "0  Varies (∼23,000 in total)            ∼10         Varies  \n",
      "1                         56             ∼4            286  \n",
      "2                      19674              4          78637  \n",
      "3  Varies (∼20,500 in total)            ∼10         Varies  \n",
      "4                      19114              4          76441  \n",
      "5                      16977            3-4  17,032 arrays  \n",
      "**************TABLE 2\n",
      "  .mw-parser-output .navbar{display:inline;font-size:88%;font-weight:normal}.mw-parser-output .navbar-collapse{float:left;text-align:left}.mw-parser-output .navbar-boxtext{word-spacing:0}.mw-parser-output .navbar ul{display:inline-block;white-space:nowrap;line-height:inherit}.mw-parser-output .navbar-brackets::before{margin-right:-0.125em;content:\"[ \"}.mw-parser-output .navbar-brackets::after{margin-left:-0.125em;content:\" ]\"}.mw-parser-output .navbar li{word-spacing:-0.125em}.mw-parser-output .navbar-mini abbr{font-variant:small-caps;border-bottom:none;text-decoration:none;cursor:inherit}.mw-parser-output .navbar-ct-full{font-size:114%;margin:0 7em}.mw-parser-output .navbar-ct-mini{font-size:114%;margin:0 4em}.mw-parser-output .infobox .navbar{font-size:100%}.mw-parser-output .navbox .navbar{display:block;font-size:100%}.mw-parser-output .navbox-title .navbar{float:left;text-align:left;margin-right:0.5em}vteScience Breakthroughs of the Year  \\\n",
      "0                                     Sciencejournal                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                          \n",
      "\n",
      "  .mw-parser-output .navbar{display:inline;font-size:88%;font-weight:normal}.mw-parser-output .navbar-collapse{float:left;text-align:left}.mw-parser-output .navbar-boxtext{word-spacing:0}.mw-parser-output .navbar ul{display:inline-block;white-space:nowrap;line-height:inherit}.mw-parser-output .navbar-brackets::before{margin-right:-0.125em;content:\"[ \"}.mw-parser-output .navbar-brackets::after{margin-left:-0.125em;content:\" ]\"}.mw-parser-output .navbar li{word-spacing:-0.125em}.mw-parser-output .navbar-mini abbr{font-variant:small-caps;border-bottom:none;text-decoration:none;cursor:inherit}.mw-parser-output .navbar-ct-full{font-size:114%;margin:0 7em}.mw-parser-output .navbar-ct-mini{font-size:114%;margin:0 4em}.mw-parser-output .infobox .navbar{font-size:100%}.mw-parser-output .navbox .navbar{display:block;font-size:100%}.mw-parser-output .navbox-title .navbar{float:left;text-align:left;margin-right:0.5em}vteScience Breakthroughs of the Year.1  \n",
      "0  1996: HIV understanding 1997: Dolly the sheep ...                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                           \n",
      "**************TABLE 3\n",
      "                   0                                        1\n",
      "0  Authority control  BNF: cb17091624w (data) GND: 1116100770\n"
     ]
    }
   ],
   "source": [
    "for index,table in enumerate(tables):\n",
    "    print(\"**************TABLE {}\".format(index))\n",
    "    print(table)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Library</th>\n",
       "      <th>ID</th>\n",
       "      <th>Species</th>\n",
       "      <th>PI</th>\n",
       "      <th>Genes targeted</th>\n",
       "      <th>gRNAs per gene</th>\n",
       "      <th>Total gRNAs</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Bassik Mouse CRISPR Knockout Library</td>\n",
       "      <td>1000000121 — 1000000130</td>\n",
       "      <td>Mouse</td>\n",
       "      <td>Bassik</td>\n",
       "      <td>Varies (∼23,000 in total)</td>\n",
       "      <td>∼10</td>\n",
       "      <td>Varies</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Mouse Tumor Suppressor Gene CRISPR Knockout Li...</td>\n",
       "      <td>113584 EFS backbone 113585 TBG backbone</td>\n",
       "      <td>Mouse</td>\n",
       "      <td>Chen</td>\n",
       "      <td>56</td>\n",
       "      <td>∼4</td>\n",
       "      <td>286</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Brie mouse genome-wide library</td>\n",
       "      <td>73632 (1 plasmid) 73633 (2 plasmid)</td>\n",
       "      <td>Mouse</td>\n",
       "      <td>Doench and Root</td>\n",
       "      <td>19674</td>\n",
       "      <td>4</td>\n",
       "      <td>78637</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Bassik Human CRISPR Knockout Library</td>\n",
       "      <td>101926 — 101934</td>\n",
       "      <td>Human</td>\n",
       "      <td>Bassik</td>\n",
       "      <td>Varies (∼20,500 in total)</td>\n",
       "      <td>∼10</td>\n",
       "      <td>Varies</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Brunello human genome-wide library</td>\n",
       "      <td>73179 (1 plasmid) 73178 (2 plasmid)</td>\n",
       "      <td>Human</td>\n",
       "      <td>Doench and Root</td>\n",
       "      <td>19114</td>\n",
       "      <td>4</td>\n",
       "      <td>76441</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>Mini-human AsCpf1-based Human Genome-wide Knoc...</td>\n",
       "      <td>130630</td>\n",
       "      <td>Human</td>\n",
       "      <td>Draetta</td>\n",
       "      <td>16977</td>\n",
       "      <td>3-4</td>\n",
       "      <td>17,032 arrays</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                             Library  \\\n",
       "0               Bassik Mouse CRISPR Knockout Library   \n",
       "1  Mouse Tumor Suppressor Gene CRISPR Knockout Li...   \n",
       "2                     Brie mouse genome-wide library   \n",
       "3               Bassik Human CRISPR Knockout Library   \n",
       "4                 Brunello human genome-wide library   \n",
       "5  Mini-human AsCpf1-based Human Genome-wide Knoc...   \n",
       "\n",
       "                                        ID Species               PI  \\\n",
       "0                  1000000121 — 1000000130   Mouse           Bassik   \n",
       "1  113584 EFS backbone 113585 TBG backbone   Mouse             Chen   \n",
       "2      73632 (1 plasmid) 73633 (2 plasmid)   Mouse  Doench and Root   \n",
       "3                          101926 — 101934   Human           Bassik   \n",
       "4      73179 (1 plasmid) 73178 (2 plasmid)   Human  Doench and Root   \n",
       "5                                   130630   Human          Draetta   \n",
       "\n",
       "              Genes targeted gRNAs per gene    Total gRNAs  \n",
       "0  Varies (∼23,000 in total)            ∼10         Varies  \n",
       "1                         56             ∼4            286  \n",
       "2                      19674              4          78637  \n",
       "3  Varies (∼20,500 in total)            ∼10         Varies  \n",
       "4                      19114              4          76441  \n",
       "5                      16977            3-4  17,032 arrays  "
      ]
     },
     "execution_count": 35,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = tables[1]\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style  type=\"text/css\" >\n",
       "    #T_c27cd21e_9052_11eb_aab7_0ed564d06023row0_col0 {\n",
       "            color:  black;\n",
       "        }    #T_c27cd21e_9052_11eb_aab7_0ed564d06023row0_col1 {\n",
       "            color:  black;\n",
       "        }    #T_c27cd21e_9052_11eb_aab7_0ed564d06023row0_col2 {\n",
       "            color:  black;\n",
       "        }    #T_c27cd21e_9052_11eb_aab7_0ed564d06023row0_col3 {\n",
       "            color:  black;\n",
       "        }    #T_c27cd21e_9052_11eb_aab7_0ed564d06023row0_col4 {\n",
       "            color:  black;\n",
       "        }    #T_c27cd21e_9052_11eb_aab7_0ed564d06023row0_col5 {\n",
       "            color:  black;\n",
       "        }    #T_c27cd21e_9052_11eb_aab7_0ed564d06023row0_col6 {\n",
       "            color:  black;\n",
       "        }    #T_c27cd21e_9052_11eb_aab7_0ed564d06023row1_col0 {\n",
       "            color:  black;\n",
       "        }    #T_c27cd21e_9052_11eb_aab7_0ed564d06023row1_col1 {\n",
       "            color:  black;\n",
       "        }    #T_c27cd21e_9052_11eb_aab7_0ed564d06023row1_col2 {\n",
       "            color:  black;\n",
       "        }    #T_c27cd21e_9052_11eb_aab7_0ed564d06023row1_col3 {\n",
       "            color:  black;\n",
       "        }    #T_c27cd21e_9052_11eb_aab7_0ed564d06023row1_col4 {\n",
       "            color:  black;\n",
       "        }    #T_c27cd21e_9052_11eb_aab7_0ed564d06023row1_col5 {\n",
       "            color:  black;\n",
       "        }    #T_c27cd21e_9052_11eb_aab7_0ed564d06023row1_col6 {\n",
       "            color:  black;\n",
       "        }    #T_c27cd21e_9052_11eb_aab7_0ed564d06023row2_col0 {\n",
       "            color:  black;\n",
       "        }    #T_c27cd21e_9052_11eb_aab7_0ed564d06023row2_col1 {\n",
       "            color:  black;\n",
       "        }    #T_c27cd21e_9052_11eb_aab7_0ed564d06023row2_col2 {\n",
       "            color:  black;\n",
       "        }    #T_c27cd21e_9052_11eb_aab7_0ed564d06023row2_col3 {\n",
       "            color:  black;\n",
       "        }    #T_c27cd21e_9052_11eb_aab7_0ed564d06023row2_col4 {\n",
       "            color:  black;\n",
       "        }    #T_c27cd21e_9052_11eb_aab7_0ed564d06023row2_col5 {\n",
       "            color:  black;\n",
       "        }    #T_c27cd21e_9052_11eb_aab7_0ed564d06023row2_col6 {\n",
       "            color:  black;\n",
       "        }    #T_c27cd21e_9052_11eb_aab7_0ed564d06023row3_col0 {\n",
       "            color:  black;\n",
       "        }    #T_c27cd21e_9052_11eb_aab7_0ed564d06023row3_col1 {\n",
       "            color:  black;\n",
       "        }    #T_c27cd21e_9052_11eb_aab7_0ed564d06023row3_col2 {\n",
       "            color:  red;\n",
       "        }    #T_c27cd21e_9052_11eb_aab7_0ed564d06023row3_col3 {\n",
       "            color:  black;\n",
       "        }    #T_c27cd21e_9052_11eb_aab7_0ed564d06023row3_col4 {\n",
       "            color:  black;\n",
       "        }    #T_c27cd21e_9052_11eb_aab7_0ed564d06023row3_col5 {\n",
       "            color:  black;\n",
       "        }    #T_c27cd21e_9052_11eb_aab7_0ed564d06023row3_col6 {\n",
       "            color:  black;\n",
       "        }    #T_c27cd21e_9052_11eb_aab7_0ed564d06023row4_col0 {\n",
       "            color:  black;\n",
       "        }    #T_c27cd21e_9052_11eb_aab7_0ed564d06023row4_col1 {\n",
       "            color:  black;\n",
       "        }    #T_c27cd21e_9052_11eb_aab7_0ed564d06023row4_col2 {\n",
       "            color:  red;\n",
       "        }    #T_c27cd21e_9052_11eb_aab7_0ed564d06023row4_col3 {\n",
       "            color:  black;\n",
       "        }    #T_c27cd21e_9052_11eb_aab7_0ed564d06023row4_col4 {\n",
       "            color:  black;\n",
       "        }    #T_c27cd21e_9052_11eb_aab7_0ed564d06023row4_col5 {\n",
       "            color:  black;\n",
       "        }    #T_c27cd21e_9052_11eb_aab7_0ed564d06023row4_col6 {\n",
       "            color:  black;\n",
       "        }    #T_c27cd21e_9052_11eb_aab7_0ed564d06023row5_col0 {\n",
       "            color:  black;\n",
       "        }    #T_c27cd21e_9052_11eb_aab7_0ed564d06023row5_col1 {\n",
       "            color:  black;\n",
       "        }    #T_c27cd21e_9052_11eb_aab7_0ed564d06023row5_col2 {\n",
       "            color:  red;\n",
       "        }    #T_c27cd21e_9052_11eb_aab7_0ed564d06023row5_col3 {\n",
       "            color:  black;\n",
       "        }    #T_c27cd21e_9052_11eb_aab7_0ed564d06023row5_col4 {\n",
       "            color:  black;\n",
       "        }    #T_c27cd21e_9052_11eb_aab7_0ed564d06023row5_col5 {\n",
       "            color:  black;\n",
       "        }    #T_c27cd21e_9052_11eb_aab7_0ed564d06023row5_col6 {\n",
       "            color:  black;\n",
       "        }</style><table id=\"T_c27cd21e_9052_11eb_aab7_0ed564d06023\" ><thead>    <tr>        <th class=\"blank level0\" ></th>        <th class=\"col_heading level0 col0\" >Library</th>        <th class=\"col_heading level0 col1\" >ID</th>        <th class=\"col_heading level0 col2\" >Species</th>        <th class=\"col_heading level0 col3\" >PI</th>        <th class=\"col_heading level0 col4\" >Genes targeted</th>        <th class=\"col_heading level0 col5\" >gRNAs per gene</th>        <th class=\"col_heading level0 col6\" >Total gRNAs</th>    </tr></thead><tbody>\n",
       "                <tr>\n",
       "                        <th id=\"T_c27cd21e_9052_11eb_aab7_0ed564d06023level0_row0\" class=\"row_heading level0 row0\" >0</th>\n",
       "                        <td id=\"T_c27cd21e_9052_11eb_aab7_0ed564d06023row0_col0\" class=\"data row0 col0\" >Bassik Mouse CRISPR Knockout Library</td>\n",
       "                        <td id=\"T_c27cd21e_9052_11eb_aab7_0ed564d06023row0_col1\" class=\"data row0 col1\" >1000000121 — 1000000130</td>\n",
       "                        <td id=\"T_c27cd21e_9052_11eb_aab7_0ed564d06023row0_col2\" class=\"data row0 col2\" >Mouse</td>\n",
       "                        <td id=\"T_c27cd21e_9052_11eb_aab7_0ed564d06023row0_col3\" class=\"data row0 col3\" >Bassik</td>\n",
       "                        <td id=\"T_c27cd21e_9052_11eb_aab7_0ed564d06023row0_col4\" class=\"data row0 col4\" >Varies (∼23,000 in total)</td>\n",
       "                        <td id=\"T_c27cd21e_9052_11eb_aab7_0ed564d06023row0_col5\" class=\"data row0 col5\" >∼10</td>\n",
       "                        <td id=\"T_c27cd21e_9052_11eb_aab7_0ed564d06023row0_col6\" class=\"data row0 col6\" >Varies</td>\n",
       "            </tr>\n",
       "            <tr>\n",
       "                        <th id=\"T_c27cd21e_9052_11eb_aab7_0ed564d06023level0_row1\" class=\"row_heading level0 row1\" >1</th>\n",
       "                        <td id=\"T_c27cd21e_9052_11eb_aab7_0ed564d06023row1_col0\" class=\"data row1 col0\" >Mouse Tumor Suppressor Gene CRISPR Knockout Library</td>\n",
       "                        <td id=\"T_c27cd21e_9052_11eb_aab7_0ed564d06023row1_col1\" class=\"data row1 col1\" >113584 EFS backbone 113585 TBG backbone</td>\n",
       "                        <td id=\"T_c27cd21e_9052_11eb_aab7_0ed564d06023row1_col2\" class=\"data row1 col2\" >Mouse</td>\n",
       "                        <td id=\"T_c27cd21e_9052_11eb_aab7_0ed564d06023row1_col3\" class=\"data row1 col3\" >Chen</td>\n",
       "                        <td id=\"T_c27cd21e_9052_11eb_aab7_0ed564d06023row1_col4\" class=\"data row1 col4\" >56</td>\n",
       "                        <td id=\"T_c27cd21e_9052_11eb_aab7_0ed564d06023row1_col5\" class=\"data row1 col5\" >∼4</td>\n",
       "                        <td id=\"T_c27cd21e_9052_11eb_aab7_0ed564d06023row1_col6\" class=\"data row1 col6\" >286</td>\n",
       "            </tr>\n",
       "            <tr>\n",
       "                        <th id=\"T_c27cd21e_9052_11eb_aab7_0ed564d06023level0_row2\" class=\"row_heading level0 row2\" >2</th>\n",
       "                        <td id=\"T_c27cd21e_9052_11eb_aab7_0ed564d06023row2_col0\" class=\"data row2 col0\" >Brie mouse genome-wide library</td>\n",
       "                        <td id=\"T_c27cd21e_9052_11eb_aab7_0ed564d06023row2_col1\" class=\"data row2 col1\" >73632 (1 plasmid) 73633 (2 plasmid)</td>\n",
       "                        <td id=\"T_c27cd21e_9052_11eb_aab7_0ed564d06023row2_col2\" class=\"data row2 col2\" >Mouse</td>\n",
       "                        <td id=\"T_c27cd21e_9052_11eb_aab7_0ed564d06023row2_col3\" class=\"data row2 col3\" >Doench and Root</td>\n",
       "                        <td id=\"T_c27cd21e_9052_11eb_aab7_0ed564d06023row2_col4\" class=\"data row2 col4\" >19674</td>\n",
       "                        <td id=\"T_c27cd21e_9052_11eb_aab7_0ed564d06023row2_col5\" class=\"data row2 col5\" >4</td>\n",
       "                        <td id=\"T_c27cd21e_9052_11eb_aab7_0ed564d06023row2_col6\" class=\"data row2 col6\" >78637</td>\n",
       "            </tr>\n",
       "            <tr>\n",
       "                        <th id=\"T_c27cd21e_9052_11eb_aab7_0ed564d06023level0_row3\" class=\"row_heading level0 row3\" >3</th>\n",
       "                        <td id=\"T_c27cd21e_9052_11eb_aab7_0ed564d06023row3_col0\" class=\"data row3 col0\" >Bassik Human CRISPR Knockout Library</td>\n",
       "                        <td id=\"T_c27cd21e_9052_11eb_aab7_0ed564d06023row3_col1\" class=\"data row3 col1\" >101926 — 101934</td>\n",
       "                        <td id=\"T_c27cd21e_9052_11eb_aab7_0ed564d06023row3_col2\" class=\"data row3 col2\" >Human</td>\n",
       "                        <td id=\"T_c27cd21e_9052_11eb_aab7_0ed564d06023row3_col3\" class=\"data row3 col3\" >Bassik</td>\n",
       "                        <td id=\"T_c27cd21e_9052_11eb_aab7_0ed564d06023row3_col4\" class=\"data row3 col4\" >Varies (∼20,500 in total)</td>\n",
       "                        <td id=\"T_c27cd21e_9052_11eb_aab7_0ed564d06023row3_col5\" class=\"data row3 col5\" >∼10</td>\n",
       "                        <td id=\"T_c27cd21e_9052_11eb_aab7_0ed564d06023row3_col6\" class=\"data row3 col6\" >Varies</td>\n",
       "            </tr>\n",
       "            <tr>\n",
       "                        <th id=\"T_c27cd21e_9052_11eb_aab7_0ed564d06023level0_row4\" class=\"row_heading level0 row4\" >4</th>\n",
       "                        <td id=\"T_c27cd21e_9052_11eb_aab7_0ed564d06023row4_col0\" class=\"data row4 col0\" >Brunello human genome-wide library</td>\n",
       "                        <td id=\"T_c27cd21e_9052_11eb_aab7_0ed564d06023row4_col1\" class=\"data row4 col1\" >73179 (1 plasmid) 73178 (2 plasmid)</td>\n",
       "                        <td id=\"T_c27cd21e_9052_11eb_aab7_0ed564d06023row4_col2\" class=\"data row4 col2\" >Human</td>\n",
       "                        <td id=\"T_c27cd21e_9052_11eb_aab7_0ed564d06023row4_col3\" class=\"data row4 col3\" >Doench and Root</td>\n",
       "                        <td id=\"T_c27cd21e_9052_11eb_aab7_0ed564d06023row4_col4\" class=\"data row4 col4\" >19114</td>\n",
       "                        <td id=\"T_c27cd21e_9052_11eb_aab7_0ed564d06023row4_col5\" class=\"data row4 col5\" >4</td>\n",
       "                        <td id=\"T_c27cd21e_9052_11eb_aab7_0ed564d06023row4_col6\" class=\"data row4 col6\" >76441</td>\n",
       "            </tr>\n",
       "            <tr>\n",
       "                        <th id=\"T_c27cd21e_9052_11eb_aab7_0ed564d06023level0_row5\" class=\"row_heading level0 row5\" >5</th>\n",
       "                        <td id=\"T_c27cd21e_9052_11eb_aab7_0ed564d06023row5_col0\" class=\"data row5 col0\" >Mini-human AsCpf1-based Human Genome-wide Knockout Library</td>\n",
       "                        <td id=\"T_c27cd21e_9052_11eb_aab7_0ed564d06023row5_col1\" class=\"data row5 col1\" >130630</td>\n",
       "                        <td id=\"T_c27cd21e_9052_11eb_aab7_0ed564d06023row5_col2\" class=\"data row5 col2\" >Human</td>\n",
       "                        <td id=\"T_c27cd21e_9052_11eb_aab7_0ed564d06023row5_col3\" class=\"data row5 col3\" >Draetta</td>\n",
       "                        <td id=\"T_c27cd21e_9052_11eb_aab7_0ed564d06023row5_col4\" class=\"data row5 col4\" >16977</td>\n",
       "                        <td id=\"T_c27cd21e_9052_11eb_aab7_0ed564d06023row5_col5\" class=\"data row5 col5\" >3-4</td>\n",
       "                        <td id=\"T_c27cd21e_9052_11eb_aab7_0ed564d06023row5_col6\" class=\"data row5 col6\" >17,032 arrays</td>\n",
       "            </tr>\n",
       "    </tbody></table>"
      ],
      "text/plain": [
       "<pandas.io.formats.style.Styler at 0x7f7f9bfe9c88>"
      ]
     },
     "execution_count": 36,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def color(val):\n",
    "    color = 'red' if val =='Human' else 'black'\n",
    "    return 'color: %s' % color\n",
    "\n",
    "s = df.style.applymap(color)\n",
    "s"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "---\n",
    "\n",
    "## Submitting Your Work\n",
    "\n",
    "In order to submit your work, you'll need to use the `git` command line program to **add** your homework file (this file) to your local repository, **commit** your changes to your local repository, and then **push** those changes up to github.com.  From there, I'll be able to **pull** the changes down and do my grading.  I'll provide some feedback, **commit** and **push** my comments back to you.  Next week, I'll show you how to **pull** down my comments.\n",
    "\n",
    "To run through everything one last time and submit your work:\n",
    "1. Use the `Kernel` -> `Restart Kernel and Run All Cells` menu option to run everything from top to bottom and stop here.\n",
    "2. Save this note with Ctrl-S (or Cmd-S)\n",
    "2. Skip down to the last command cell (the one starting with `%%bash`) and run that cell.\n",
    "\n",
    "If anything fails along the way with this submission part of the process, let me know.  I'll help you troubleshoort."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "assert False, \"DO NOT REMOVE THIS LINE\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "---"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Already up to date.\n",
      "[main 549fb78] Submitting the week 8 programming assignment\n",
      " 2 files changed, 935 insertions(+), 52 deletions(-)\n",
      " create mode 100644 week08/week08_assignment_2.ipynb\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "remote: This repository moved. Please use the new location:        \n",
      "remote:   git@github.com:Xiaoniu86/hds5210-2021.git        \n",
      "To github.com:xiaoniu86/hds5210-2021.git\n",
      "   c8ab4cb..549fb78  main -> main\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "git pull\n",
    "git add week08_assignment_2.ipynb\n",
    "git commit -a -m \"Submitting the week 8 programming assignment\"\n",
    "git push"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "---\n",
    "\n",
    "If the message above says something like _Submitting the week 8 programming assignment_ or _Everything is up to date_, then your work was submitted correctly."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
